### PR TITLE
Delete otherlib/{graph,win32graph}

### PR DESCRIPTION
**Pull request by ghost**
This PR deletes the graphics library. The plan is to distribute it as a separate package starting from OCaml 4.09.

I prepared a separate repository to be new home of the graphics library: https://github.com/ocaml/graphics. The git history was preserved in the process. The package is fully functional and ready to be released once 4.09 is released.

Users of the graphics library won't be affected by this change as long as they rely on the "graphics" findlib package. I checked all the users of graphics in opam, and found only one that will be affected by this PR: ocapic. Ocapic will need to be updated in order to be compatible with 4.09.

### CODE REVIEW

**Comment by nojb** [comment]
I think you need to update the variable `otherlibs_all` in the main `Makefile` as well.

File: configure
Comment (nojb) [inline configure]: This part of the diff seems unrelated to the PR.

**Comment by ghost** [comment]
Indeed. I removed it along with win32graph

File: configure
Comment (ghost) [reply configure]: Agreed. However, I didn't add it. I simply ran `autoconf` which added this stuff. Should I delete it manually? /cc @shindere

**Comment by nojb** [comment]
> Indeed. I removed it along with win32graph

Ah, sorry I had missed that.

**Comment by ghost** [comment]
I went over the reverse dependency of graphics and tried to build them against the external package (github.com/ocaml/graphics). Most of them compiled fine, I only had issues with the following packages:

- camlimages: the build was broken but the fix was a one-liner. I sent a patch to the maintainer of camlimages
- ocapic: fixing the build is a bit more complicated, I opened a ticket to discuss possible solutions
- cairo2: I couldn't try as I'm having trouble installing the C library on my machine

All the other packages compiled just fine with the external package.

So even though it is not a completely transparent transition, it looks like it is not going to cause too much trouble in practice.

**Comment by ghost** [comment]
BTW, cairo2 is using dune so either it is not affected, either the fix is trivial as for camlimages.

File: manual/manual/library/libgraph.etex
Comment (xavierleroy) [inline manual/manual/library/libgraph.etex]: The LaTeX markup could use some more care, e.g. format the URL better and make it clickable.  Perhaps @Octachron could help?  Also, it would be good to mention the OPAM package.

**Review by xavierleroy — COMMENTED** [review]
This pull request makes me all warm inside :-)  Thank you @diml.  

It looks very good to me.  @shindere should tell us if the generated `configure` is OK or if it needs to regenerated.  The manual could use some more help too.

File: manual/manual/library/libgraph.etex
Comment (Octachron) [reply manual/manual/library/libgraph.etex]: `\url{https://github.com/ocaml/graphics}` should make the link clickable.

**Comment by ghost** [comment]
You are welcome :)

I updated the manual

File: manual/manual/library/libgraph.etex
Comment (ghost) [reply manual/manual/library/libgraph.etex]: I updated the url and added a paragraph about the opam package

**Comment by shindere** [comment]
Many thanks for externalizing this. A few comments:

- The Change entry (or the commit message, but rather the change entry)
could include the URL of the new source repository as is done in the
error message in configure

- You may want to add the names of the reviewers in the Change entry

- Since Makefile.config gets installed, is it okay to just remove the
GRAPHLIB variable?

**Comment by ghost** [comment]
I updated the changelog entry. Not sure what can be done for the GRAPHLIB variable. We could set it do a dummy value that we know will break. @dra27 what do you think, is it better to just remove it or set it to a dummy value such as `the_graphics_library_is_no_longer_distributed_with_ocaml`?

BTW, I'm a bit lost regarding what's happening with the `configure` file: initially, an extra block about `runstate` was generated. Then On Thursday I rebased my branch, ran `autoconf` again the extra block was no longer generated: all good. Now, this morning I rebased again my branch, ran `autoconf` again and the diff shows that the `runstate` block is being removed by my PR (???). I'm guessing this is related to the version of autoconf being used by the person generating the configure script...

**Comment by shindere** [comment]
Yeah, I think it has to do with the autoconf version used to generate
the script. Perhaps, since we don't really need this block we could make
sure this does not get committed. At the same time I'd say it does not
really hurt. Also, to me that's the problem you gen when storing
generated files in the repo. I'm not saying `configure` shouldn't be
distributed in the tarballs of the compiler, quite the contrary
actually, I do think it should be distributed. But I think it'd just be
better to not have it in the repo, or at least not on trunk. It oculd be
added e.g. to the release branches.

**Comment by ghost** [comment]
That seems right to me as well. I have always been allergic to committing such files in repositories.

**Comment by shindere** [comment]
Jérémie Dimino (2019/03/18 05:07 -0700):
> That seems right to me as well. I have always been allergic to
> committing such files in repositories.

OK shall I submit a PR that removes configure from trunk then? At least
so that this can be discussed? If I do so I'd like to Cc you @diml if
you don't mind.

**Comment by dra27** [comment]
For autoconf - the `runstate` thing is because there is a very old patch which is used by virtually *all* distros (and I think is committed to autoconf) but which has never been released. It usually disappears when you run autoconf on Cygwin and, given their usual philosophies, on the BSDs as well, I think.

For not committing `configure` - I have in mind to sort a patch which both declares `configure` as binary (which stops it from being included in diff views) and also add a test on Travis to ensure that `configure.ac` generates `configure` as given in the commit. I may manage to get to that this week, who knows!

`configure` tends to be an irritating special case for committing - I agree that it shouldn't need to be - though it is convenient not to require autotools in order to build a Git checkout of OCaml.

**Comment by shindere** [comment]
Well, to me the travis test and all this sounds a bit overkill, I have
to say.

I would way more prefer to not have configure at all in the repo.
I feel it would be both easier and less fragile to addapt to this than
to declare it binary,make sure it corresponds to configure.ac (which
will require making sure the two versions of autoconf on the dev's box
and on Travis generate the smae file.)

**Comment by ghost** [comment]
Yh, from the contributor's point of view, it's much easier if the configure script is not committed. Many people are not root on their machine at work so installing a particular version of autoconf might be a lot of hassle.

I vote for the configure script to not be committed. @shindere happy to be cced.

**Comment by ghost** [comment]
BTW, from my point of view this PR is ready to be merged, just waiting for approval :)

**Comment by alainfrisch** [comment]
> Many people are not root on their machine at work so installing a particular version of autoconf might be a lot of hassle.

Are you making the assumption that one good-enough version of autoconf is readily available, for all developers and people that want to compile non-released code?  With a committed `configure`, perhaps it's a more tedious to contribute for people who need to touch the configure script, but it's easier for all other users who simply want to compile or contribute to other parts.

**Comment by alainfrisch** [comment]
Also, I'm concerned that people could start using autoconf features which are not widely available.  We might need to document which version is expected, not only for maintainers, but for everyone who will compile the code from Git.

**Comment by dra27** [comment]
No - "virtually everyone" uses autoconf 2.69 with the runstate patch. autoconf 2.69 was released in 2012.

**Comment by dra27** [comment]
@diml - your preference is contradictory. If `configure` is not committed then all contributors need autoconf because you then can't build OCaml without it?

**Comment by xavierleroy** [comment]
>  "virtually everyone" uses autoconf 2.69 with the runstate patch.

And Debian uses autoconf 2.69 without the patch, correct?

If that's true, I don't think we have much of a problem: when reviewing changes that affect configure.ac, either we don't look at the changes in generated configure, or we have  a quick look knowing that the runstate stuff is unimportant.

For the record, I prefer to have configure committed in trunk and in release branches.  Among other reasons, it's good to have exactly the same build procedures for trunk and release branches.

**Comment by dra27** [comment]
Ubuntu uses it with the patch, so I'd assumed Debian does - I've not checked, though.

**Comment by dra27** [comment]
Debian does have - it patch 6 in https://sources.debian.org/src/autoconf/2.69-10/debian/patches/series/

**Comment by xavierleroy** [comment]
OK, I read the diff the wrong way.  So, for some reason the autoconf used by @diml doesn't have the runstate patch, or did not generate the runstate block.  But "virtually everyone" should not run into this issue.  For this reason I propose that:
* @diml re-runs autoconf on various machines until the generated configure looks right;
* everyone else stops worrying about committing configure.

**Comment by ghost** [comment]
Ok, so the autoconf on my laptop does generate the runstate block. I regenerated it and pushed. I was actually the one who removed the runstate block in the PR to remove vmthreads...

**Review by xavierleroy — APPROVED** [review]
Looks good, thanks.  I fixed a documentation typo myself.  Will merge next.

**Comment by shindere** [comment]
Alain Frisch (2019/03/18 06:21 -0700):
> Also, I'm concerned that people could start using autoconf features
> which are not widely available.  We might need to document which
> version is expected, not only for maintainers, but for everyone who
> will compile the code from Git.

Should we want to do such a thing, we could use the `AC_PREREQ` macro to
enforce (rather than just document) an autoconf minimal version to be
used.

**Comment by shindere** [comment]
Xavier Leroy (2019/03/18 13:42 +0000):
> For the record, I prefer to have configure committed in trunk and in
> release branches.  Among other reasons, it's good to have exactly the
> same build procedures for trunk and release branches.

Well then I'd prefer to have configure commited nowhere and just be part
of the tarballs, but I understand this is not going to happen so I'm
writing htis just for the record as well.

**Comment by shindere** [comment]
Xavier Leroy (2019/03/18 06:51 -0700):
> OK, I read the diff the wrong way.  So, for some reason the autoconf used by @diml doesn't have the runstate patch, or did not generate the runstate block.  But "virtually everyone" should not run into this issue.  For this reason I propose that:
> * @diml re-runs autoconf on various machines until the generated configure looks right;
> * everyone else stops worrying about committing configure.

Well commiting does create worries, whether we like it or not! ;-)

**Pull request merged**
